### PR TITLE
Windows: Support loading WIM images

### PR DESCRIPTION
**Pull request by jstarks**
Windows base layers are distributed by Microsoft in the WIM file format.
This change allows Docker to read them natively so that it can load
Windows base layers directly.

To preserve tar as the format of choice throughout Docker, this change
works by converting WIM files to tar streams on demand.

Currently this blocks docker save and docker push for images that use a
Docker-managed Windows base layer. This is because there is no equivalent
to tar-split for WIMs. Fixing this would probably require storing the WIM
separately.

This also adds a small tool that takes a WIM and produces a tar file
suitable for passing to docker load. This is currently the only way to
ingest WIM files.

Signed-off-by: John Starks jostarks@microsoft.com

@aaronlehmann I think this, combined with an ability of images to reference layers by URL, will allow us to support docker pull of Windows base images. Does this match what you were thinking? Any progress on the "foreign" layer support?

cc/@jhowardmsft

### CODE REVIEW

File: layer/custom_unix.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// +build !windows
```
Comment (jhowardmsft) [inline layer/custom_unix.go]: Minor, but most _unix files have //+build linux freebsd

File: pkg/wimutils/iswim.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"bytes"
+	"encoding/binary"
+	"io"
+)
+
+var wimMagic = [8]byte{'M', 'S', 'W', 'I', 'M'}
+
+// IsWIM looks at the first 12 bytes of a file to determine if the underlying stream
+// is a WIM file.
+func IsWIM(r io.ReaderAt) (bool, error) {
```
Comment (jhowardmsft) [inline pkg/wimutils/iswim.go]: Should there be an iswim_windows.go and iswim_unix.go, where the unix version always returns false/error, as there's no WIM handling on Linux?

File: pkg/wimutils/tar_windows.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func writeTarFromImage(w io.Writer, img *wim.Image, uimg *wim.Image) error {
+	root, err := img.Open()
+	if err != nil {
+		return err
+	}
+	tw := tar.NewWriter(w)
+
+	// For now, put WIM contents under a directory called Files, since
```
Comment (jhowardmsft) [inline pkg/wimutils/tar_windows.go]: Should this have a // TODO Windows: Post TP5....

**Comment by jhowardmsft** [comment]
Nice! Minor comments.

File: pkg/wimutils/iswim.go (hunk 2 above)
Comment (jstarks) [inline pkg/wimutils/iswim.go]: It's still a WIM, even if you're on Linux. :)

Comment (jhowardmsft) [inline pkg/wimutils/iswim.go]: Fair :+1:

**Comment by tonistiigi** [comment]
Being able to load/pull and not save/push an image later is quite breaking to the current docker ux. If this is only for the technical reasons then I think we should find a workaround(probably store wim on disk). If it also has some legal issues then it's more complicated.

@aaronlehmann Maybe the format detection would make more sense on the layerstore level? The `ReaderAt` could be replaced with `Peek()`.

**Comment by aaronlehmann** [comment]
I agree that I'd like to see a plan for how to support pushing base layers. Since the WIM is converted to a tar internally, I wonder if it's possible to store tar-split metadata for this converted tar, and then preserve the ability to push the base layer in tar format. This would also involve hashing the tar stream to determine a layer ID instead of hashing the WIM directly. But I don't know if the converted tar is somehow machine-specific.

As for legal issues, my recollection from the last time we talked about this is that it should be okay to provide the capability to push base layers, as long as we are careful about UI and make it difficult to do accidentally (i.e. don't push bsae layers by default, and add an appropriate set of flags/prompts to make sure the user really wants to push the base layer and is aware of the copyright implications). I think it is important to preserve the technical ability to push or export a base layer even if the typical use case won't involve doing it.

I haven't had a chance to work on "foreign" layer support. I imagine the way this would work is that Docker associates with each base layer the URL where it can be publicly downloaded, and the cryptographic hash of the file at that URL. Then, when it creates a manifest for an image and doesn't actually push the base layer as part of it, it includes that source URL and the hash instead of referencing a blob that's local to the registry. Is that what you have in mind?

**Comment by jstarks** [comment]
The converted tar is not machine-specific, so that tar (or its tar-split + contents) could be used as a base layer on any machine. The trouble is that the construction of this tar may not be byte-for-byte deterministic, since future changes to the tar package or wim package may result in semantically equivalent but different tar streams. So if I understand the content addressability requirements, this will be problematic -- the layer ID would be different on different machines if the machine got the layer purely in WIM format.

I guess in theory we could support a model where the public repository distributes just the tar-split portion but not the file contents, and provides the URL of the WIM to use to get the contents. This would mean that when a new WIM is available, we (Microsoft) would sideload the WIM in some way locally and then push it to the public repo, which would push just the tar-split and the WIM URL. If a user wanted to push to a private repo for an air-gapped scenario, perhaps we could allow them to push a full tar and not just the tar-split data.

I'm not sure if this meets the legal requirements. @taylorb-microsoft FYI.

**Comment by jstarks** [comment]
@tonistiigi WIM files cannot be streamed -- they are inherently a random access format. So even though Peek() could be used to detect a WIM, it can't be used to read a WIM. We need ReaderAt to read the contents, which means we need to do the reading before the file gets passed through the various helper Reader wrappers (progress.Reader, etc.).

**Comment by aaronlehmann** [comment]
Would it be possible for Microsoft to distribute the base layer preconverted to tar format? That way the engine would not need to support a second format, and the content-addressability requirements would be fulfilled by tar-split.

**Comment by jstarks** [comment]
@aaronlehmann I don't think we'll realistically be able to distribute tar images of these, at least for a while.

**Comment by thaJeztah** [comment]
CI is failing;

```
02:04:40 # github.com/docker/docker/contrib/wim2img
02:04:40 contrib/wim2img/wim2img.go:91: undefined: wimutils.TarFromWIM
02:04:40 contrib/wim2img/wim2img.go:165: undefined: image.FeatureWin32k
02:05:33 # github.com/docker/docker/distribution/xfer
02:05:33 distribution/xfer/download_test.go:213: cannot use mockDownloadDescriptor literal (type *mockDownloadDescriptor) as type DownloadDescriptor in array element:
.....
```

**Comment by jstarks** [comment]
@thaJeztah Thanks. I'm going to hold off making more changes to this until we can determine what design changes should be made.

**Comment by jstarks** [comment]
We're going to distribute tar files natively.

**Pull request closed without merging**